File: 프로그래머스/기초 트레이닝 - Level 0/정수를 나선형으로 배치하기-solve.java
class Solution {
    public int[][] solution(int n) {
        int[][] answer = new int[n][n];
        int[][] moves = new int[][]{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int num = 2, end = n * n; 
        int x = 0, y = 0, dir = 0;
        answer[x][y] = 1;
        
        while (num <= end) {
            int nx = x + moves[dir][0], ny = y + moves[dir][1];
            
            if (0 <= nx && nx < n && 0 <= ny && ny < n && answer[nx][ny] == 0) {
                answer[nx][ny] = num++;
                x = nx; y = ny; 
            } else {
                dir = (dir + 1) % moves.length;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/기초 트레이닝 - Level 0/정수를 나선형으로 배치하기-solve.py
def solution(n):
    answer = [[0] * n for _ in range(n)]
    moves = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    num, end = 2, n * n 
    x, y, dir = 0, 0, 0
    answer[x][y] = 1
    
    while num <= end: 
        nx, ny = x + moves[dir][0], y + moves[dir][1]
        
        if 0 <= nx < n and 0 <= ny < n and answer[nx][ny] == 0:
            answer[nx][ny] = num
            num += 1
            x, y = nx, ny
        else:
            dir = (dir + 1) % len(moves)
        
    return answer